### DISCUSSION TITLE
Node.js binary cannot be executed

### DISCUSSION THREAD

**Discussion by winkloid** [nextcloud/recognize#1009] [Q&A] [answered]
Hey there, I am not able to get Recognize to work with the LinuxServer.io nextcloud docker image. I'm only getting the notification that the node.js binary could not be found/executed. Even if I manually set the path to the node.js binary which is delivered with recognize inside the directory "/config/www/nextcloud/apps/recognize/bin/node"  I get the same notification. Using the command line with the user that is used to execute the container (UID 100) I was able to find out that the permissions needed to execute the node binary are set:
```
/config/www/nextcloud/apps/recognize/bin$ ls -la
total 105144
drwxr-xr-x  2 abc users     4096 Oct 19 20:11 .
drwxr-xr-x 12 abc users     4096 Oct 19 20:11 ..
-rw-r--r--  1 abc users 34142344 Oct 19 20:11 arm64.tar.gz
-rwxr-xr-x  1 abc users 73509256 Oct 19 20:11 node

```
I also tried different binaries that I downloaded from the node.js Website but not a single solved the problem.

When running `occ recognize:classify` I get the following output when recognize tries to recognize faces:
```
Running array (
  0 => '/bin/nice',
  1 => '-0',
  2 => '/config/www/nextcloud/apps/recognize/bin/node',
  3 => '/config/www/nextcloud/apps/recognize/src/classifier_faces.js',
  4 => '-',
)
Classifier process output: nice: ‘/config/www/nextcloud/apps/recognize/bin/node’: No such file or directory

Classifier process output: nice: ‘/config/www/nextcloud/apps/recognize/bin/node’: No such file or directory

Error while running facesclassifier
No files left to classify
No files left to classify
No files left to classify
```

### Details about my system
The Docker Container is running on a ARM64 vServer with 4GB RAM and 6GB swap memory. The vServer is hosted by Hetzner (CAX11).

Maybe someone faced a similar problem and is able to help :)

**Comment by winkloid** [comment, marked as answer]
In the meantime I was able to found a solution for this problem: The problem was that Alpine Linux, on whose base image the Nextcloud LSIO image is based, works with musl instead of glibc. As a result, the binaries that can be downloaded from the Node.JS website by default cannot be executed and the aforementioned error is output. My solution was to put a binary on the system that can also be used under Alpine Linux with musl. The obvious solution was to use the Alpine package manager, with which I installed a compatible node.JS version: `apk add nodejs`. The binary can now be found under `/usr/bin/node`. After setting this path in my Nextcloud Admin Panel for Recognize the error is gone:
![grafik](https://github.com/nextcloud/recognize/assets/73366117/d48f2501-dbd4-41d6-bbf1-5b9253150a93)

Note that this error will probably appear again after every update of the nextcloud container if no volume is mounted that contains the node binary file.

**Reply by SayantanRC** [reply]
My host system isn't alpine linux. How do I get this binary? @winkloid any idea?

Edit: For anyone who is a noob - you have to open the nextcloud docker container shell (`docker exec -it <container id> bash`). Then install nodejs as `apk add nodejs` and set `/usr/bin/node` in Recognize. Now here is where I faced some problem. After setting the path, Recognize still kept showing the error. All you have to do is open any other nextcloud page and the come back to Recognize settings and it'll be green.

**Reply by relink2013** [reply]
> Note that this error will probably appear again after every update of the nextcloud container if no volume is mounted that contains the node binary file.

With that install method, yes you'll need to re-install nodejs after each update of the container. However all LSIO containers have a permanent way to install packages.

You only need to add two variables to your container:
```
DOCKER_MODS=linuxserver/mods:universal-package-install
```
```
INSTALL_PACKAGES=nodejs
```

You can read more about that specific feature [here](https://github.com/linuxserver/docker-mods/tree/universal-package-install). [Here](https://mods.linuxserver.io) for a list of other mods they have for their containers. Plus theres more [here](https://docs.linuxserver.io/general/container-customization/) that can be customized.

**Reply by SayantanRC** [reply]
@relink2013 are these to be added under the `environment` section in the docker compose file?
